.github/workflows/nubots.yaml

```
@@ -0,0 +1,171 @@
+# Continuous Integration tests
+name: NUbots CI
+
+# Controls when the action will run.
+on:
+  # Triggers the workflow on push or pull request events but only for the main branch
+  push:
+    branches: [main]
+  pull_request:
+    branches: [main]
+
+# A workflow run is made up of one or more jobs that can run sequentially or in parallel
+jobs:
+  # Build the docker image
+  build_docker:
+    name: "Build docker image"
+
+    # The type of runner that the job will run on
+    runs-on: ubuntu-20.04
+
+    # We output the image tag that we create so the other jobs can use it
+    outputs:
+      image: ${{ steps.image_output.outputs.image }}
+
+    # Steps represent a sequence of tasks that will be executed as part of the job
+    steps:
+      # Cancels previously running jobs for this workflow
+      - name: Cancel Workflow Action
+        uses: styfle/cancel-workflow-action@0.9.1
+
+      # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
+      - name: Checkout Code
+        uses: actions/checkout@v2
+
+      # Setup docker buildx
+      - name: 🐳 Set up docker buildx 🐳
+        uses: docker/setup-buildx-action@v1
+
+      - name: Login to DockerHub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_PASSWORD }}
+
+      # Build the docker image
+      - name: 🐳 Build the docker image 🐳
+        uses: docker/build-push-action@v2
+        with:
+          pull: true
+          tags: "nubots/nubots:${{ hashFiles('docker/**') }}"
```

Can we just use the branch name here instead of a hash?
```suggestion
          tags: "nubots/nubots:${{ GITHUB_REF#refs/heads/ }}"
```

---

Even better, use the PR number. But the idea behind this was that you only choose a different image if the actual files in the docker folder change (not just if it's a different branch). In theory two branches could have the same docker files and therefore share an image.

The biggest problem with any of this is cleaning up the images once we are done with the PR

---

Ah ok makes sense, yeah PR number would be better. With the hash I dont think different PR's with the same dockerfiles could share images because of `cache-from: type=registry,ref=nubots/nubots:generic` ?

---

Actually nvm, It would cache the common layers between the generic and hash image because of the `pull: true` I believe.

---

I have done something similar. It probably won't work though :laughing:

---

That is the idea, that it will always use the common layers in the nubots:generic (so if you don't change the image like 99% of PRs it'll just entirely cache)